services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/UploadDirectoryConfiguration.java

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.transfer.s3;
+
+import java.util.Objects;
+import java.util.Optional;
+import software.amazon.awssdk.annotations.SdkPreviewApi;
+import software.amazon.awssdk.annotations.SdkPublicApi;
+import software.amazon.awssdk.utils.ToString;
+import software.amazon.awssdk.utils.builder.CopyableBuilder;
+import software.amazon.awssdk.utils.builder.ToCopyableBuilder;
+
+/**
+ * Configuration options for {@link S3TransferManager#uploadDirectory}. All values are optional, and not specifying them will
+ * use the SDK default values.
+ *
+ * <p>Use {@link #builder()} to create a set of options.</p>
+ */
+@SdkPublicApi
+@SdkPreviewApi
+public final class UploadDirectoryConfiguration implements ToCopyableBuilder<UploadDirectoryConfiguration.Builder,
+    UploadDirectoryConfiguration> {
+
+    private final Boolean followSymbolicLinks;
+    private final Integer maxDepth;
+    private final Boolean recursive;
+
+    public UploadDirectoryConfiguration(DefaultBuilder builder) {
+        this.followSymbolicLinks = builder.followSymbolicLinks;
+        this.maxDepth = builder.maxDepth;
+        this.recursive = builder.recursive;
+    }
+
+    /**
+     * @return whether to follow symbolic links
+     */
+    public Optional<Boolean> followSymbolicLinks() {
```

Comment:
>  because it's possible that the caller knows whether or not they're present:

The caller doesn't know though and would have to do null check explicitly otherwise.

https://github.com/aws/aws-sdk-java-v2/blob/c47d69e5f96885cb7f5efa868950a94a5ed73b5a/services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/TransferConfiguration.java#L42-L45

> We're inconsistent already, so we should at least make a choice.

Hmm, I think ClientOverrideConfiguartion and RequestOverrideConfiguration are consistent though; all getters are returning optional values if the values are nullable. Did I miss anything?